### PR TITLE
Add Category to provide filename or translated filename

### PR DESCRIPTION
**Pull request by ghost**
Re: https://github.com/ResearchKit/ResearchKit/pull/324
Re: https://github.com/ResearchKit/ResearchKit/issues/323

Add a method in a category on ORKFileResult to provide a filename derived from the concatenation of the ORKResult identifier and ORKStep identifier. 
Provide for translation of the derived filename via FilenameTranslation.json.

This would serve our purposes for recovering the original filenames in Parkinson and Cardio without adding the step identifier back into the logName in ORKRecorder.

### CODE REVIEW

File: APCAppCore/APCAppCore/Library/Categories/ORKFileResult+Filename.m [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+#import "ORKFileResult+Filename.h"
+
+static NSString *kJSONExtension = @"json";
+static NSString *kTranslationFilename = @"FilenameTranslation";
+
+@implementation ORKFileResult (Filename)
+
++ (NSString *)filenameForFileResultIdentifier: (NSString *)fileResultIdentifier stepIdentifier: (NSString *)stepIdentifier
+{
+    NSString *identifier = [[fileResultIdentifier stringByAppendingString:@"_"]stringByAppendingString:stepIdentifier];
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Library/Categories/ORKFileResult+Filename.m]: Need a space before `stringBy...`

**Comment by YuanZhu-apple** [comment]
I assume `FilenameTranslation.json` and the method caller will be in the apps?

File: APCAppCore/APCAppCore/Library/Categories/ORKFileResult+Filename.m [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    
+    NSString *JSONString = [[NSString alloc] initWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:NULL];
+    NSError *parseError;
+    
+    NSDictionary *translationDictionary = [NSJSONSerialization JSONObjectWithData:[JSONString dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableContainers error:&parseError];
+    
+    if ([translationDictionary objectForKey:identifier]) {
+        translatedFilename = [translationDictionary objectForKey:identifier];
+    }else{
+        translatedFilename = identifier;
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Library/Categories/ORKFileResult+Filename.m]: Avoid calling `objectForKey:identifier` twice. How about this?

```
translatedFilename = [translationDictionary objectForKey:identifier];
if (translatedFilename == nil) {
    translatedFilename = identifier;
}
```

or 

```
translatedFilename = [translationDictionary objectForKey:identifier] ? : identifier;
```

**Comment by ghost** [comment]
@YuanZhu-apple  Yes, as needed. Sage has a filename schema which we need to follow, else they will have fragmented data in their database. This will also help insulate the app from further changes to the url which we are currently using to identify the file and form a filename.

File: APCAppCore/APCAppCore/Library/Categories/ORKFileResult+Filename.m [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+#import "ORKFileResult+Filename.h"
+
+static NSString *kJSONExtension = @"json";
+static NSString *kTranslationFilename = @"FilenameTranslation";
+
+@implementation ORKFileResult (Filename)
+
++ (NSString *)filenameForFileResultIdentifier: (NSString *)fileResultIdentifier stepIdentifier: (NSString *)stepIdentifier
+{
+    NSString *identifier = [[fileResultIdentifier stringByAppendingString:@"_"] stringByAppendingString:stepIdentifier];
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Library/Categories/ORKFileResult+Filename.m:47]: What if `fileResultIdentifier` is nil?

**Comment by ghost** [comment]
@YuanZhu-apple  To clarify further on your question about the caller. 
Currently we have an APCDataArchiver class which includes a friendlyFilenameForFile() method for filename translation via regex. I am currently refactoring / deprecating this class by moving the Parkinson specific code into subclasses of APCBaseTaskViewController and using my encapsulated upload classes to upload. The filename translation will be initially be called from APCDataArchiver, and after refactoring from APCBaseTaskViewController.

**Comment by YuanZhu-apple** [comment]
r=me.

**Pull request merged**